// col-assets/lib/api.js
/**
 * Copyright 2015 UC Berkeley (UCB) Licensed under the
 * Educational Community License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may
 * obtain a copy of the License at
 *
 *     http://opensource.org/licenses/ECL-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an "AS IS"
 * BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

var _ = require('lodash');
var Joi = require('joi');

var ActivitiesAPI = require('col-activities');
var CategoriesAPI = require('col-categories');
var CollabosphereConstants = require('col-core/lib/constants');
var CollabosphereUtil = require('col-core/lib/util');
var DB = require('col-core/lib/db');
var log = require('col-core/lib/logger')('col-assets');
var UserConstants = require('col-users/lib/constants');

/**
 * Get an asset
 *
 * @param  {Context}        ctx                             Standard context containing the current user and the current course
 * @param  {Number}         id                              The id of the asset
 * @param  {Boolean}        [incrementViews]                Whether the total number of views for the asset should be incremented by 1
 * @param  {Function}       callback                        Standard callback function
 * @param  {Object}         callback.err                    An error that occurred, if any
 * @param  {Asset}          callback.asset                  The requested asset
 */
var getAsset = module.exports.getAsset = function(ctx, id, incrementViews, callback) {
  // Parameter validation
  var validationSchema = Joi.object().keys({
    'id': Joi.number().required(),
    'incrementViews': Joi.boolean().optional()
  });

  var validationResult = Joi.validate({
    'id': id,
    'incrementViews': incrementViews
  }, validationSchema);

  if (validationResult.error) {
    return callback({'code': 400, 'msg': validationResult.error.details[0].message});
  }

  // Get the asset from the DB
  var options = {
    'where': {
      'id': id,
      'course_id': ctx.course.id
    },
    'include': [
      {
        'model': DB.Activity,
        'attributes': ['type'],
        'required': false,
        'where': {
          'course_id': ctx.course.id,
          'user_id': ctx.user.id,
          'type': ['like', 'dislike'],
          'asset_id': id,
          'object_type': 'asset'
        }
      },
      {
        'model': DB.User,
        'attributes': UserConstants.BASIC_USER_FIELDS
      }, {
        'model': DB.Category
      }, {
        'model': DB.Comment,
        'include': [{
          'model': DB.User,
          'attributes': UserConstants.BASIC_USER_FIELDS
        }]
      }
    ]
  };
  DB.Asset.find(options).complete(function(err, asset) {
    if (err) {
      log.error({'err': err, 'id': id}, 'Failed to get an asset');
      return callback({'code': 500, 'msg': err.message});
    } else if (!asset) {
      log.debug({'err': err, 'id': id}, 'An asset with the specified id could not be found');
      return callback({'code': 404, 'msg': 'An asset with the specified id could not be found'});
    }

    // Increment the number of views for the asset if specified
    if (incrementViews) {
      asset.increment('views').complete(function(err, asset) {
        if (err) {
          log.error({'err': err, 'id': id}, 'Failed to increase the views on an asset');
          return callback({'code': 500, 'msg': err.message});
        }

        return callback(null, asset);
      });
    } else {
      return callback(null, asset);
    }
  });
};

/**
 * Get the assets for the current course
 *
 * @param  {Context}        ctx                             Standard context containing the current user and the current course
 * @param  {Number}         [limit]                         The maximum number of results to retrieve. Defaults to 10
 * @param  {Number}         [offset]                        The number to start paging from. Defaults to 0
 * @param  {Function}       callback                        Standard callback function
 * @param  {Object}         callback.err                    An error that occurred, if any
 * @param  {Object}         callback.assets                 The retrieved assets
 * @param  {Number}         callback.assets.total           The total number of assets in the current course
 * @param  {Asset[]}        callback.assets.results         The paged assets in the current course
 */
var getAssets = module.exports.getAssets = function(ctx, limit, offset, callback) {
  // Default some parameters
  limit = CollabosphereUtil.getNumberParam(limit, 10, 1, 25);
  offset = CollabosphereUtil.getNumberParam(offset, 0, 0);

  // Parameter validation
  var validationSchema = Joi.object().keys({
    'limit': Joi.number().required(),
    'offset': Joi.number().required()
  });

  var validationResult = Joi.validate({
    'limit': limit,
    'offset': offset
  }, validationSchema);

  if (validationResult.error) {
    return callback({'code': 400, 'msg': validationResult.error.details[0].message});
  }

  // Get the assets from the DB
  var options = {
    'where': {
      'course_id': ctx.course.id
    },
    'order': [['created_at', 'DESC']],
    'limit': limit,
    'offset': offset,
    'include': [
      {
        'model': DB.Activity,
        'attributes': ['type'],
        'required': false,
        'where': {
          'course_id': ctx.course.id,
          'user_id': ctx.user.id,
          'type': ['like', 'dislike'],
          'object_type': 'asset'
        }
      },
      {
        'model': DB.User,
        'attributes': UserConstants.BASIC_USER_FIELDS
      }
    ]
  };
  DB.Asset.findAndCountAll(options).complete(function(err, result) {
    if (err) {
      log.error({'err': err, 'course': ctx.course}, 'Failed to get the assets in the current course');
      return callback({'code': 500, 'msg': err.message});
    }

    return callback(null, {
      'total': result.count,
      'results': result.rows
    });
  });
};

/**
 * Create a new link asset
 *
 * @param  {Context}        ctx                             Standard context containing the current user and the current course
 * @param  {String}         title                           The title of the link
 * @param  {String}         url                             The url of the link
 * @param  {Object}         [opts]                          A set of optional parameters
 * @param  {Number[]}       [opts.categories]               The ids of the categories to which the link should be associated
 * @param  {String}         [opts.description]              The description of the link
 * @param  {String}         [opts.source]                   The source of the link
 * @param  {Function}       callback                        Standard callback function
 * @param  {Object}         callback.err                    An error that occurred, if any
 * @param  {Asset}          callback.asset                  The created link asset
 */
var createLink = module.exports.createLink = function(ctx, title, url, opts, callback) {
  // Default the title to the provided url
  title = title || url;

  // Default the optional values
  opts = opts || {};
  opts.categories = opts.categories || [];

  // Parameter validation
  var validationSchema = Joi.object().keys({
    'title': Joi.string().max(255).required(),
    'url': Joi.string().uri().max(255).required(),
    'opts': Joi.object().keys({
      'categories': Joi.array().unique().items(Joi.number()).optional(),
      'description': Joi.string().allow('').optional(),
      'source': Joi.string().uri().optional()
    })
  });

  var validationResult = Joi.validate({
    'title': title,
    'url': url,
    'opts': opts
  }, validationSchema);

  if (validationResult.error) {
    return callback({'code': 400, 'msg': validationResult.error.details[0].message});
  }

  // Ensure that all provided categories exist
  CategoriesAPI.getCategoriesById(ctx, opts.categories, function(err, categories) {
    if (err) {
      return callback(err);
    }

    // Create the link asset in the DB
    var asset = {
      'course_id': ctx.course.id,
      'user_id': ctx.user.id,
      'type': 'link',
      'title': title,
      'url': url,
      'description': opts.description,
      'source': opts.source
    };

    // TODO: Remove this once proper thumbnail functionality is in place
    var imageFormats = ['jpg', 'jpeg', 'png'];
    if (_.contains(imageFormats, url.toLowerCase().split('.').pop())) {
      asset.thumbnail_url = url;
    }

    DB.Asset.create(asset).complete(function(err, asset) {
      if (err) {
        log.error({'err': err}, 'Failed to create a new link asset');
        return callback({'code': 500, 'msg': err.message});
      }

      // Add the associated categories to the asset
      asset.setCategories(opts.categories).complete(function(err) {
        if (err) {
          log.error({'err': err}, 'Failed to associate categories to a new link asset');
          return callback({'code': 500, 'msg': err.message});
        }

        // Give the user points for adding a new asset
        ActivitiesAPI.createActivity(ctx.course, ctx.user, 'add_asset', asset.id, CollabosphereConstants.ACTIVITY.OBJECT_TYPES.ASSET, null, function(err) {
          if (err) {
            return callback(err);
          }

          // Retrieve the created asset, including the associated user and
          // associated categories
          getAsset(ctx, asset.id, false, callback);
        });
      });
    });
  });
};

/**
 * Edit an asset
 *
 * @param  {Context}        ctx                             Standard context containing the current user and the current course
 * @param  {Number}         id                              The id of the asset that is being edited
 * @param  {String}         title                           The updated title of the asset
 * @param  {Object}         [opts]                          A set of optional parameters
 * @param  {Number[]}       [opts.categories]               The updated ids of the categories to which the asset should be associated
 * @param  {String}         [opts.description]              The updated description of the asset
 * @param  {Object}         callback.err                    An error that occurred, if any
 * @param  {Asset}          callback.asset                  The updated asset
 */
var editAsset = module.exports.editAsset = function(ctx, id, title, opts, callback) {
  opts = opts || {};
  opts.categories = opts.categories || [];
  opts.description = opts.description || '';

  // Parameter validation
  var validationSchema = Joi.object().keys({
    'id': Joi.number().required(),
    'title': Joi.string().max(255).required(),
    'opts': Joi.object().keys({
      'categories': Joi.array().unique().items(Joi.number()).optional(),
      'description': Joi.string().allow('').optional()
    })
  });

  var validationResult = Joi.validate({
    'id': id,
    'title': title,
    'opts': opts
  }, validationSchema);

  if (validationResult.error) {
    return callback({'code': 400, 'msg': validationResult.error.details[0].message});
  }

  // Ensure that the provided asset exists
  getAsset(ctx, id, false, function(err, asset) {
    if (err) {
      return callback(err);
    }

    // Only course administrators and the user that has created the asset are allowed
    // to edit it
    if (!ctx.user.is_admin && asset.user_id !== ctx.user.id) {
      log.error({'course': ctx.course}, 'Unauthorized to edit an asset');
      return callback({'code': 401, 'msg': 'Unauthorized to edit an asset'});
    }

    // Ensure that all provided categories exist
    CategoriesAPI.getCategoriesById(ctx, opts.categories, function(err, categories) {
      if (err) {
        return callback(err);
      }

      // Update the asset in the DB
      var update = {
        'title': title,
        'description': opts.description
      };

      asset.updateAttributes(update).complete(function(err, asset) {
        if (err) {
          log.error({'err': err, 'id': id}, 'Failed to update an asset');
          return callback({'code': 500, 'msg': err.message});
        }

        // Update the associated categories to the asset
        asset.setCategories(opts.categories).complete(function(err) {
          if (err) {
            log.error({'err': err}, 'Failed to update the categories of an asset');
            return callback({'code': 500, 'msg': err.message});
          }

          // Retrieve the created asset, including the associated user and
          // associated categories
          getAsset(ctx, asset.id, false, callback);
        });
      });
    });
  });
};

/* COMMENTS */

/**
 * Get a comment.
 * Note that this is a private method that doesn't do any validation
 *
 * @param  {Number}         id                              The id of the comment to retrieve
 * @param  {Object}         callback.err                    An error that occurred, if any
 * @param  {Comment}        callback.comment                The created comment
 * @param  {User}           callback.comment.user           The user that created the comment
 * @api private
 */
var getComment = function(id, callback) {
  var options = {
    'where': {
      'id': id
    },
    'include': [{
      'model': DB.User,
      'attributes': UserConstants.BASIC_USER_FIELDS
    }]
  };
  DB.Comment.find(options).complete(function(err, comment) {
    if (err) {
      log.error({'err': err, 'id': id}, 'Failed to retrieve the created comment');
      return callback({'code': 500, 'msg': err.message});
    }

    return callback(null, comment);
  });
};

/**
 * Create a new comment on an asset
 *
 * @param  {Context}        ctx                             Standard context containing the current user and the current course
 * @param  {Number}         assetId                         The id of the asset to which the comment is added
 * @param  {String}         body                            The body of the comment
 * @param  {Number}         [parent]                        The id of the comment to which the comment is a reply
 * @param  {Function}       callback                        Standard callback function
 * @param  {Object}         callback.err                    An error that occurred, if any
 * @param  {Comment}        callback.comment                The created comment
 * @param  {User}           callback.comment.user           The user that created the comment
 */
var createComment = module.exports.createComment = function(ctx, assetId, body, parent, callback) {
  // Parameter validation
  var validationSchema = Joi.object().keys({
    'assetId': Joi.number().required(),
    'body': Joi.string().required(),
    'parent': Joi.number().optional()
  });

  var validationResult = Joi.validate({
    'assetId': assetId,
    'body': body,
    'parent': parent
  }, validationSchema);

  if (validationResult.error) {
    return callback({'code': 400, 'msg': validationResult.error.details[0].message});
  }

  // Get the asset to which the comment is being added
  getAsset(ctx, assetId, false, function(err, asset) {
    if (err) {
      return callback(err);
    }

    // Check that the comment to which this comment is a reply exists
    if (parent) {
      // Ensure the parent id is a number
      parent = CollabosphereUtil.getNumberParam(parent);
      var parentComment = _.findWhere(asset.toJSON().comments, {'id': parent});
      if (!parentComment) {
        log.error({'err': err, 'parent': parent}, 'Failed to find the comment to which a reply is being made');
        return callback({'code': 400, 'msg': 'Failed to find the comment to which a reply is being made'});
      }
    }

    // Create the link asset in the database
    var comment = {
      'asset_id': assetId,
      'user_id': ctx.user.id,
      'parent_id': parent,
      'body': body
    };

    DB.Comment.create(comment).complete(function(err, comment) {
      if (err) {
        log.error({'err': err}, 'Failed to create a new comment');
        return callback({'code': 500, 'msg': err.message});
      }

      // Increase the comment count
      asset.increment('comment_count').complete(function(err, asset) {

        // Retrieve the created comment, including the associated user
        getComment(comment.id, callback);
      });
    });
  });
};

/**
 * Edit a comment on an asset
 *
 * @param  {Context}        ctx                             Standard context containing the current user and the current course
 * @param  {Number}         assetId                         The id of the asset to which the comment belongs
 * @param  {Number}         id                              The id of the comment that is being edited
 * @param  {String}         body                            The updated comment body
 * @param  {Function}       callback                        Standard callback function
 * @param  {Object}         callback.err                    An error that occurred, if any
 * @param  {Comment}        callback.comment                The updated comment
 * @param  {User}           callback.comment.user           The user that created the comment
 */
var editComment = module.exports.editComment = function(ctx, assetId, id, body, callback) {
  // Parameter validation
  var validationSchema = Joi.object().keys({
    'assetId': Joi.number().required(),
    'id': Joi.number().required(),
    'body': Joi.string().required()
  });

  var validationResult = Joi.validate({
    'assetId': assetId,
    'id': id,
    'body': body
  }, validationSchema);

  if (validationResult.error) {
    return callback({'code': 400, 'msg': validationResult.error.details[0].message});
  }

  // Get the comment that is being edited
  var options = {
    'where': {
      'id': id,
      'asset_id': assetId
    },
    'include': [DB.Asset]
  };

  DB.Comment.find(options).complete(function(err, comment) {
    if (err) {
      log.error({'err': err, 'id': id, 'assetId': assetId}, 'Failed to retrieve the comment to update');
      return callback({'code': 500, 'msg': err.message});
    } else if (!comment) {
      log.debug({'err': err, 'id': id, 'assetId': assetId}, 'The comment that is being edited could not be found');
      return callback({'code': 404, 'msg': 'The comment that is being edited could not be found'});
    }

    // Verify that the user is allowed to edit the comment. A user is
    // only able to edit a comment when the user manages the course or
    // the comment is their own
    if ((!ctx.user.is_admin || comment.asset.course_id !== ctx.course.id) && comment.user_id !== ctx.user.id) {
      log.error({'assetId': assetId, 'id': id}, 'Unauthorized to edit the comment');
      return callback({'code': 401, 'msg': 'Unauthorized to edit the comment'});
    }

    // Update the comment
    var update = {
      'body': body
    };

    comment.updateAttributes(update).complete(function(err, comment) {
      if (err) {
        log.error({'err': err}, 'Failed to update a comment');
        return callback({'code': 500, 'msg': err.message});
      }

      // Retrieve the updated comment, including the associated user
      getComment(comment.id, callback);
    });
  });
};

/**
 * Delete a comment on an asset
 *
 * @param  {Context}        ctx                             Standard context containing the current user and the current course
 * @param  {Number}         assetId                         The id of the asset to which the comment belongs
 * @param  {Number}         id                              The id of the comment that is being deleted
 * @param  {Function}       callback                        Standard callback function
 * @param  {Object}         callback.err                    An error that occurred, if any
 */
var deleteComment = module.exports.deleteComment = function(ctx, assetId, id, callback) {
  // Parameter validation
  var validationSchema = Joi.object().keys({
    'assetId': Joi.number().required(),
    'id': Joi.number().required()
  });

  var validationResult = Joi.validate({
    'assetId': assetId,
    'id': id
  }, validationSchema);

  if (validationResult.error) {
    return callback({'code': 400, 'msg': validationResult.error.details[0].message});
  }

  // Get the comment that is being deleted
  var options = {
    'where': {
      'id': id,
      'asset_id': assetId
    },
    'include': [DB.Asset]
  };

  DB.Comment.find(options).complete(function(err, comment) {
    if (err) {
      log.error({'err': err, 'id': id, 'assetId': assetId}, 'Failed to retrieve the comment to delete');
      return callback({'code': 500, 'msg': err.message});
    } else if (!comment) {
      log.debug({'err': err, 'id': id, 'assetId': assetId}, 'The comment that is being deleted could not be found');
      return callback({'code': 404, 'msg': 'The comment that is being deleted could not be found'});
    }

    // Verify that the user is allowed to delete the comment. A user is
    // only able to delete a comment when the user manages the course or
    // the comment is their own
    if ((!ctx.user.is_admin || comment.asset.course_id !== ctx.course.id) && comment.user_id !== ctx.user.id) {
      log.error({'assetId': assetId, 'id': id}, 'Unauthorized to delete the comment');
      return callback({'code': 401, 'msg': 'Unauthorized to delete the comment'});
    }

    // Verify that the comment doesn't have any replies. A comment can not
    // be removed if it has any existing replies
    options = {
      'where': {
        'parent_id': id,
        'asset_id': assetId
      }
    };

    DB.Comment.findAll(options).complete(function(err, comments) {
      if (err) {
        log.error({'err': err, 'id': id, 'assetId': assetId}, 'Failed to retrieve the replies of the comment to delete');
        return callback({'code': 500, 'msg': err.message});
      }

      // The comment can not be deleted if it has any replies
      if (comments.length > 0) {
        log.error({'assetId': assetId, 'id': id}, 'The comment can not be deleted as it has replies');
        return callback({'code': 400, 'msg': 'The comment can not be deleted as it has replies'});
      }

      comment.destroy().complete(function(err) {
        if (err) {
          log.error({'err': err, 'category': category.id}, 'Failed to delete a comment');
          return callback({'code': 500, 'msg': err.message});
        }

        // Decrease the comment count
        comment.asset.decrement('comment_count').complete(function(err, asset) {

          // TODO: Delete the points for the comment

          return callback();
        });
      });
    });
  });
};

/* (DIS)LIKING */

/**
 * Like or dislike an asset
 *
 * @param  {Context}        ctx                             Standard context containing the current user and the current course
 * @param  {Number}         assetId                         The id of the asset that is liked or disliked
 * @param  {Boolean}        [like]                          `true` when the asset should be liked, `false` when the asset should be disliked. When `null` is provided, the previous like or dislike will be undone
 * @param  {Function}       callback                        Standard callback function
 * @param  {Object}         callback.err                    An error that occurred, if any
 */
var like = module.exports.like = function(ctx, assetId, like, callback) {
  // Parameter validation
  var validationSchema = Joi.object().keys({
    'assetId': Joi.number().required(),
    'like': Joi.boolean().optional()
  });

  var validationResult = Joi.validate({
    'assetId': assetId,
    'like': like
  }, validationSchema);

  if (validationResult.error) {
    return callback({'code': 400, 'msg': validationResult.error.details[0].message});
  }

  // Get the asset that is being liked or disliked
  getAsset(ctx, assetId, false, function(err, asset) {
    if (err) {
      return callback(err);
    }

    // Users are not able to like or dislike their own assets
    if (asset.user_id === ctx.user.id) {
      return callback({'code': 401, 'msg': 'You can not like or dislike your own assets'});
    }

    // Remove the previous like or dislike
    ActivitiesAPI.deleteActivity(ctx.course, ctx.user, ['like', 'dislike'], asset.id, CollabosphereConstants.ACTIVITY.OBJECT_TYPES.ASSET, null, function(err, deletedLike) {
      if (err) {
        return callback(err);
      }

      // Remove the previous get like or get dislike
      ActivitiesAPI.deleteActivity(ctx.course, asset.user, ['get_like', 'get_dislike'], asset.id, CollabosphereConstants.ACTIVITY.OBJECT_TYPES.ASSET, ctx.user, function(err) {
        if (err) {
          return callback(err);
        }

        // Update the like and dislike count on the asset
        updateAssetLikeCount(asset, like, deletedLike, function(err, asset) {
          if (err) {
            return callback(err);
          }

          // If a like or dislike is being undone, no further actions are required
          if (!_.isBoolean(like)) {
            return callback();
          }

          // Add the like / dislike activity
          var type = 'dislike';
          if (like) {
            type = 'like';
          }

          ActivitiesAPI.createActivity(ctx.course, ctx.user, type, asset.id, CollabosphereConstants.ACTIVITY.OBJECT_TYPES.ASSET, null, function(err) {
            if (err) {
              return callback(err);
            }

            // Add the get like / get dislike activity
            type = 'get_dislike';
            if (like) {
              type = 'get_like';
            }

            ActivitiesAPI.createActivity(ctx.course, asset.user, type, asset.id, CollabosphereConstants.ACTIVITY.OBJECT_TYPES.ASSET, ctx.user, function(err) {
              if (err) {
                return callback(err);
              }

              return callback();
            });
          });
        });
      });
    });
  });
};

/**
 * Adjust the cached like and dislike count on an asset after removing a previous like or dislike
 *
 * @param  {Asset}          asset                           The asset for which to update the like and dislike count
 * @param  {Boolean}        like                            `true` when the asset should be liked, `false` when the asset should be disliked. When `null` is provided, the previous like or dislike will be undone
 * @param  {Activity}       deletedLike                     The removed like or dislike activity on the asset
 * @param  {Function}       callback                        Standard callback function
 * @param  {Object}         callback.err                    An error that occurred, if any
 * @param  {Asset}          callback.asset                  The updated asset
 * @api private
 */
var updateAssetLikeCount = function(asset, like, deletedLike, callback) {
  var likeDelta = 0;
  var dislikeDelta = 0;
  if (deletedLike) {
    if (deletedLike.type === 'like') {
      likeDelta--;
    } else {
      dislikeDelta--;
    }
  }
  if (like === true) {
    likeDelta++;
  } else if (like === false) {
    dislikeDelta++;
  }

  asset.increment({'likes': likeDelta, 'dislikes': dislikeDelta}).complete(function(err, asset) {
    if (err) {
      log.error({'err': err, 'asset': asset}, 'Failed to update the like and dislike count on an asset');
      return callback({'code': 500, 'msg': err.message});
    }

    return callback(null, asset);
  });
};
