// col-categories/tests/util.js
/**
 * Copyright 2015 UC Berkeley (UCB) Licensed under the
 * Educational Community License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may
 * obtain a copy of the License at
 *
 *     http://opensource.org/licenses/ECL-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an "AS IS"
 * BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

var _ = require('lodash');
var assert = require('assert');

/**
 * Assert that a category has all expected properties
 *
 * @param  {Category}           category                      The category to assert the properties for
 * @param  {Object}             [opts]                        Optional parameters to verify the category with
 * @param  {Category}           [opts.expectedCategory]       The category to which the provided category should be compared
 * @param  {Boolean}            [opts.expectAssetCount]       Whether the number of assets associated to the category are expected to be included
 * @throws {AssertionError}                                   Error thrown when an assertion failed
 */
var assertCategory = module.exports.assertCategory = function(category, opts) {
  opts = opts || {};

  // Ensure that all expected properties are present
  assert.ok(category);
  assert.ok(category.id);
  assert.ok(category.course_id);
  assert.ok(category.title);
  assert.ok(category.created_at);
  assert.ok(category.updated_at);

  if (category.canvas_assignment_id) {
    assert.ok(_.isFinite(category.canvas_assignment_id));
  }

  if (opts.expectAssetCount) {
    assert.ok(_.isFinite(category.asset_count));
  }

  // Ensure that all the category properties are the same as the ones for
  // the expected category
  if (opts.expectedCategory) {
    assert.strictEqual(category.id, opts.expectedCategory.id);
    assert.strictEqual(category.course_id, opts.expectedCategory.course_id);
    assert.strictEqual(category.title, opts.expectedCategory.title);
    assert.strictEqual(category.created_at, opts.expectedCategory.created_at);
    assert.strictEqual(category.updated_at, opts.expectedCategory.updated_at);

    if (opts.expectAssetCount) {
      assert.strictEqual(category.asset_count, opts.expectedCategory.asset_count);
    }

    // Ensure that all optional properties are the same as the ones for the
    // expected category
    if (category.canvas_assignment_id || opts.expectedCategory.canvas_assignment_id) {
      assert.strictEqual(category.canvas_assignment_id, opts.expectedCategory.canvas_assignment_id);
    }
  }
};

/**
 * Assert that a new category can be created
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {Course}             course                          The Canvas course in which the user is interacting with the API
 * @param  {String}             title                           The name of the category
 * @param  {Function}           callback                        Standard callback function
 * @param  {Asset}              callback.asset                  The created category
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertCreateCategory = module.exports.assertCreateCategory = function(client, course, title, callback) {
  client.categories.createCategory(course, title, function(err, category) {
    assert.ok(!err);
    assert.ok(category);
    assertCategory(category, {'expectAssetCount': true});
    assert.strictEqual(category.title, title);
    assert.ok(!category.canvas_assignment_id);

    return callback(category);
  });
};

/**
 * Assert that a new category can not be created
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {Course}             course                          The Canvas course in which the user is interacting with the API
 * @param  {String}             title                           The name of the category
 * @param  {Number}             code                            The expected HTTP error code
 * @param  {Function}           callback                        Standard callback function
 * @param  {Asset}              callback.asset                  The created category
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertCreateCategoryFails = module.exports.assertCreateCategoryFails = function(client, course, title, code, callback) {
  client.categories.createCategory(course, title, function(err, category) {
    assert.ok(err);
    assert.strictEqual(err.code, code);
    assert.ok(!category);

    return callback();
  });
};

/**
 * Assert that the categories for a course can be retrieved
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {Course}             course                          The Canvas course in which the user is interacting with the API
 * @param  {Number}             expectedTotal                   The expected total number of categories in the current course
 * @param  {Function}           callback                        Standard callback function
 * @param  {Category[]}         callback.categories             The categories in the current course
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertGetCategories = module.exports.assertGetCategories = function(client, course, expectedTotal, callback) {
  client.categories.getCategories(course, function(err, categories) {
    assert.ok(!err);
    assert.ok(categories);
    assert.strictEqual(categories.length, expectedTotal);
    _.each(categories, function(category) {
      assertCategory(category, {'expectAssetCount': true});
    });

    return callback(categories);
  });
};

/**
 * Assert that the categories for a course can not be retrieved
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {Course}             course                          The Canvas course in which the user is interacting with the API
 * @param  {Number}             code                            The expected HTTP error code
 * @param  {Function}           callback                        Standard callback function
 * @param  {Category[]}         callback.categories             The categories in the current course
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertGetCategoriesFails = module.exports.assertGetCategoriesFails = function(client, course, code, callback) {
  client.categories.getCategories(course, function(err, categories) {
    assert.ok(err);
    assert.strictEqual(err.code, code);
    assert.ok(!categories);

    return callback();
  });
};

/**
 * Assert that a category can be edited
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {Course}             course                          The Canvas course in which the user is interacting with the API
 * @param  {Number}             id                              The id of the category that is being edited
 * @param  {String}             title                           The updated category name
 * @param  {Function}           callback                        Standard callback function
 * @param  {Category}           callback.category               The updated category
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertEditCategory = module.exports.assertEditCategory = function(client, course, id, title, callback) {
  client.categories.editCategory(course, id, title, function(err, category) {
    assert.ok(!err);
    assert.ok(category);
    assertCategory(category, {'expectAssetCount': true});
    assert.strictEqual(category.id, id);
    assert.strictEqual(category.title, title);

    return callback(category);
  });
};

/**
 * Assert that a category can not be edited
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {Course}             course                          The Canvas course in which the user is interacting with the API
 * @param  {Number}             id                              The id of the category that is being edited
 * @param  {String}             title                           The updated category name
 * @param  {Number}             code                            The expected HTTP error code
 * @param  {Function}           callback                        Standard callback function
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertEditCategoryFails = module.exports.assertEditCategoryFails = function(client, course, id, title, code, callback) {
  client.categories.editCategory(course, id, title, function(err, category) {
    assert.ok(err);
    assert.strictEqual(err.code, code);
    assert.ok(!category);

    return callback();
  });
};

/**
 * Assert that a category can be deleted
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {Course}             course                          The Canvas course in which the user is interacting with the API
 * @param  {Number}             id                              The id of the category that is being deleted
 * @param  {Function}           callback                        Standard callback function
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertDeleteCategory = module.exports.assertDeleteCategory = function(client, course, id, callback) {
  client.categories.deleteCategory(course, id, function(err) {
    assert.ok(!err);

    // Verify that the category no longer exists
    client.categories.getCategories(course, function(err, categories) {
      assert.ok(!err);
      assert.ok(categories);
      assert.ok(!_.findWhere(categories, {'id': id}));

      return callback();
    });
  });
};

/**
 * Assert that a category can not be deleted
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {Course}             course                          The Canvas course in which the user is interacting with the API
 * @param  {Number}             id                              The id of the category that is being deleted
 * @param  {Number}             code                            The expected HTTP error code
 * @param  {Function}           callback                        Standard callback function
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertDeleteCategoryFails = module.exports.assertDeleteCategoryFails = function(client, course, id, code, callback) {
  client.categories.deleteCategory(course, id, function(err) {
    assert.ok(err);
    assert.strictEqual(err.code, code);

    return callback();
  });
};
