// col-whiteboards/tests/util.js
/**
 * Copyright 2015 UC Berkeley (UCB) Licensed under the
 * Educational Community License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may
 * obtain a copy of the License at
 *
 *     http://opensource.org/licenses/ECL-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an "AS IS"
 * BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

/**
 * Assert that a whiteboard has all expected properties
 *
 * @param  {Whiteboard}         whiteboard                    The whiteboard to assert the properties for
 * @param  {Object}             [opts]                        Optional parameters to verify the whiteboard with
 * @param  {Whiteboard}         [opts.expectedWhiteboard]     The whiteboard to which the provided whiteboard should be compared
 * @param  {Boolean}            [opts.expectFullWhiteboard]   Whether the full whiteboard details are expected, including the list of members, online members and whiteboard elements
 * @param  {Number}             [opts.expectedMemberCount]    The total number of members that are expected on the whiteboard
 * @throws {AssertionError}                                   Error thrown when an assertion failed
 */
var assertWhiteboard = module.exports.assertWhiteboard = function(whiteboard, opts) {
  opts = opts || {};

  // Ensure that all expected properties are present
  assert.ok(whiteboard);
  assert.ok(whiteboard.id);
  assert.ok(whiteboard.course_id);
  assert.ok(whiteboard.title);
  assert.ok(whiteboard.created_at);
  assert.ok(whiteboard.updated_at);

  // TODO: Check full properties
  if (opts.expectFullWhiteboard) {

  } else {
    assert.ok(_isFinite(whiteboard.online_count));
  }

  if (opts.expectedWhiteboard) {
    assert.strictEqual(whiteboard.id, opts.expectedAsset.id);
    assert.strictEqual(whiteboard.course_id, opts.expectedWhiteboard.course_id);
    assert.strictEqual(whiteboard.title, opts.expectedWhiteboard.title);
    assert.strictEqual(whiteboard.created_at, opts.expectedWhiteboard.created_at);
  }
};

/**
 * Assert that a new whiteboard can be created
 *
 * @param  {RestClient}         client                        The REST client to make the request with
 * @param  {Course}             course                        The Canvas course in which the user is interacting with the API
 * @param  {String}             title                         The title of the whiteboard
 * @param  {Number[]}           [members]                     The ids of the users that should be added to the whiteboard as members. The current user will automatically be added as a member
 * @param  {Function}           callback                      Standard callback function
 * @param  {Object}             callback.err                  An error that occurred, if any
 * @param  {Whiteboard}         callback.whiteboard           The created whiteboard
 * @throws {AssertionError}                                   Error thrown when an assertion failed
 */
var assertCreateWhiteboard = module.exports.assertCreateWhiteboard = function(client, course, title, members, callback) {
  client.whiteboards.createWhiteboard(course, title, members, function(err, whiteboard) {
    assert.ok(!err);
    assert.ok(whiteboard);
    assertWhiteboard(whiteboard, {'expectFullWhiteboard': true});
    assert.strictEqual(whiteboard.title, title);

    // TODO: Check members
    // _.each(members, function(member))

    return callback(whiteboard);
  });
};

/**
 * Assert that a new whiteboard can not be created
 *
 * @param  {RestClient}         client                        The REST client to make the request with
 * @param  {Course}             course                        The Canvas course in which the user is interacting with the API
 * @param  {String}             title                         The title of the whiteboard
 * @param  {Number[]}           [members]                     The ids of the users that should be added to the whiteboard as members. The current user will automatically be added as a member
 * @param  {Number}             code                          The expected HTTP error code
 * @param  {Function}           callback                      Standard callback function
 * @param  {Object}             callback.err                  An error that occurred, if any
 * @param  {Whiteboard}         callback.whiteboard           The created whiteboard
 * @throws {AssertionError}                                   Error thrown when an assertion failed
 */
var assertCreateWhiteboardFails = module.exports.assertCreateWhiteboardFails = function(client, course, title, members, code, callback) {
  client.whiteboards.createWhiteboard(course, title, members, function(err, whiteboard) {
    assert.ok(err);
    assert.strictEqual(err.code, code);
    assert.ok(!whiteboard);

    return callback();
  });
};

/**
 * Assert that a whiteboard can be retrieved
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {Course}             course                          The Canvas course in which the user is interacting with the API
 * @param  {Number}             id                              The id of the whiteboard
 * @param  {Whiteboard}         [expectedWhiteboard]            The expected whiteboard to be retrieved
 * @param  {Number}             [expectedMemberCount]           The total number of members that are expected on the whiteboard
 * @param  {Function}           callback                        Standard callback function
 * @param  {Whiteboard}         callback.whiteboard             The retrieved whiteboard
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertGetWhiteboard = module.exports.assertGetWhiteboard = function(client, course, id, expectedWhiteboard, expectedMemberCount, callback) {
  client.whiteboards.getWhiteboard(course, id, function(err, whiteboard) {
    assert.ok(!err);
    assert.ok(whiteboard);
    assert.strictEqual(whiteboard.id, id);
    assertWhiteboard(whiteboard, {'expectedWhiteboard': expectedWhiteboard, 'expectedMemberCount': expectedMemberCount});

    return callback(whiteboard);
  });
};

/**
 * Assert that a whiteboard can not be retrieved
 *
 * @param  {RestClient}         client                          The REST client to make the request with
 * @param  {Course}             course                          The Canvas course in which the user is interacting with the API
 * @param  {Number}             id                              The id of the whiteboard
 * @param  {Number}             code                            The expected HTTP error code
 * @param  {Function}           callback                        Standard callback function
 * @param  {Whiteboard}         callback.whiteboard             The retrieved whiteboard
 * @throws {AssertionError}                                     Error thrown when an assertion failed
 */
var assertGetWhiteboardFails = module.exports.assertGetWhiteboardFails = function(client, course, code, callback) {
  client.whiteboards.getWhiteboard(course, id, function(err, whiteboard) {
    assert.ok(err);
    assert.strictEqual(err.code, code);
    assert.ok(!whiteboard);

    return callback();
  });
};


var assertGetWhiteboards = module.exports.assertGetWhiteboards = function() {};
var assertGetWhiteboardsFails = module.exports.assertGetWhiteboardsFails = function() {};

var assertEditWhiteboard = module.exports.assertEditWhiteboard = function() {};
var assertEditWhiteboardFails = module.exports.assertEditWhiteboardFails = function() {};
