// col-whiteboards/tests/util.js
/**
 * Copyright 2015 UC Berkeley (UCB) Licensed under the
 * Educational Community License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may
 * obtain a copy of the License at
 *
 *     http://opensource.org/licenses/ECL-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an "AS IS"
 * BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

var assertWhiteboard = module.exports.assertWhiteboard = function(whiteboard, opts) {};

var assertCreateWhiteboard = module.exports.assertCreateWhiteboard = function() {};
var assertCreateWhiteboardFails = module.exports.assertCreateWhiteboardFails = function() {};

var assertGetWhiteboard = module.exports.assertGetWhiteboard = function() {};
var assertGetWhiteboardFails = module.exports.assertGetWhiteboardFails = function() {};

var assertGetWhiteboards = module.exports.assertGetWhiteboards = function() {};
var assertGetWhiteboardsFails = module.exports.assertGetWhiteboardsFails = function() {};

var assertEditWhiteboard = module.exports.assertEditWhiteboard = function() {};
var assertEditWhiteboardFails = module.exports.assertEditWhiteboardFails = function() {};
