// col-canvas/tests/model.js
/**
 * Copyright 2015 UC Berkeley (UCB) Licensed under the
 * Educational Community License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may
 * obtain a copy of the License at
 *
 *     http://opensource.org/licenses/ECL-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an "AS IS"
 * BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

var _ = require('lodash');

/**
 * Mock a Canvas User object
 *
 * @param {String}  name    The name of the user
 */
var CanvasUser = module.exports.CanvasUser = function(name) {
  return {
    'id': _.random(10000000),
    'name': name,
    'sortable_name': name,
    'short_name': name.split(' ')[0]
  };
};

/**
 * Mock a Canvas Assignment object
 *
 * @param  {Number}               courseId        The id of the Canvas course the assignment belongs to
 * @param  {CanvasSubmission[]}   submissions     The submissions that have been made for this assignment
 */
var CanvasAssignment = module.exports.CanvasAssignment = function(courseId, submissions) {
  return {
    'id': _.random(1000000),
    'assignment_group_id': 1,
    'automatic_peer_reviews': false,
    'created_at': '2015-05-22T13:19:16Z',
    'description': '',
    'due_at': null,
    'grade_group_students_individually': false,
    'grading_standard_id': null,
    'grading_type': 'points',
    'group_category_id': null,
    'lock_at': null,
    'peer_reviews': false,
    'points_possible': 5,
    'position': 9,
    'post_to_sis': null,
    'unlock_at': null,
    'updated_at': '2015-05-22T13:19:16Z',
    'course_id': courseId,
    'name': 'Some assignment name',
    'submission_types': ['online_text_entry', 'online_url', 'online_upload'],
    'has_submitted_submissions': !_.isEmpty(submissions),
    'muted': false,
    'html_url': 'http://localhost:3000/courses/1/assignments/10',
    'needs_grading_count': 0,
    'integration_id': null,
    'integration_data': {},
    'published': false,
    'unpublishable': true,
    'locked_for_user': false,

    // This is the only property that deviates from a regular Canvas Assignment object,
    // but it makes passing data around and mocking REST calls a bit easier if we know
    // how many submissions there are ahead of time
    'submissions': submissions
  };
};

/**
 * Mock a Canvas submission object
 *
 * @param {Number}          userId      The id of the Canvas user who made the submission
 * @param {String}          type        The type of submission. One of `online_url`, `online_text_entry` or `online_upload`
 * @param {String}          url         The user-provided URL in case of an `online_url` submission
 * @param {CanvasFile[]}    files       One or more files in case of an `online_upload` submission
 * @param {String}          text        The submitted text in case of an `online_text_entry` submission
 */
var CanvasSubmission = module.exports.CanvasSubmission = function(userId, type, url, files, text) {
  var that = {
    'assignment_id': 9,
    'attempt': 1,
    'body': null,
    'grade': null,
    'grade_matches_current_submission': true,
    'graded_at': null,
    'grader_id': null,
    'id': 19,
    'score': null,
    'submission_type': type,
    'submitted_at': '2015-02-20T22:50:01Z',
    'user_id': userId,
    'workflow_state': 'submitted',
    'late': false,
    'preview_url': 'http://localhost:3000/courses/1/assignments/9/submissions/6?preview=1'
  };

  if (type === 'online_url') {
    that.url = url;
  } else if (type === 'online_text_entry') {
    that.body = text;
  } else if (type === 'online_upload') {
    that.attachments = files;
  }
  return that;
};

/**
 * Mock a Canvas file object
 *
 * @param {String}    contentType     The mime type of the file
 * @param {String}    displayName     A name for the file as provided by the uploader
 * @param {String}    filename        The filename
 */
var CanvasFile = module.exports.CanvasFile = function(contentType, displayName, filename) {
  var id = _.random(1000000);
  return {
    'id': id,
    'content-type': contentType,
    'display_name': displayName,
    'filename': filename,
    'url': 'http://localhost:3000/files/' + id + '/download?download_frd=1&verifier=5s8rxxD8BVmVXgD188QZIaEWRxoamEopJtNOC9Sw',
    'size': _.random(100000),
    'created_at': '2015-06-09T18:00:10Z',
    'updated_at': '2015-06-09T18:00:10Z',
    'unlock_at': null,
    'locked': false,
    'hidden': false,
    'lock_at': null,
    'hidden_for_user': false,
    'thumbnail_url': 'http://localhost:3000/images/thumbnails/show/' + id,
    'locked_for_user': false,
    'preview_url': null
  };
};

/**
 * Mock a Canvas discussion object
 *
 * @param  {CanvasUser}       user      The user who created the discussion
 * @param  {CanvasEntry[]}    entries   The entries on the discussion
 */
var CanvasDiscussion = module.exports.CanvasDiscussion = function(user, entries) {
  entries = entries || [];

  var data = {
    'assignment_id': null,
    'delayed_post_at': null,
    'discussion_type': 'threaded',
    'id': _.random(1000000),
    'last_reply_at': '2015-06-08T18:51:51Z',
    'lock_at': null,
    'podcast_has_student_posts': false,
    'position': null,
    'posted_at': '2015-06-08T18:51:51Z',
    'root_topic_id': null,
    'title': 'No no',
    'user_name': 'Simon Gaeremynck',
    'discussion_subentry_count': entries.length,
    'permissions': {
      'attach': true,
      'update': true,
      'delete': true
    },
    'message': '',
    'require_initial_post': null,
    'user_can_see_posts': true,
    'podcast_url': null,
    'read_state': 'read',
    'unread_count': 0,
    'subscribed': true,
    'topic_children': [],
    'attachments': [],
    'published': false,
    'can_unpublish': true,
    'locked': false,
    'can_lock': true,
    'author': {
      'id': user.id,
      'display_name': user.fullName,
      'avatar_image_url': 'https://canvas.instructure.com/images/messages/avatar-50.png',
      'html_url': 'http://localhost:3000/courses/3/users/1'
    },
    'html_url': 'http://localhost:3000/courses/3/discussion_topics/6',
    'url': 'http://localhost:3000/courses/3/discussion_topics/6',
    'pinned': false,
    'group_category_id': null,
    'can_group': true,
    'locked_for_user': false
  };

  var that = {
    'addEntry': function(entry) {
      entries.push(entry);
      data.discussion_subentry_count = entries.length;

      if (entry.parent_id) {
        var parentEntry = _.find(entries, {'id': entry.parent_id});
        if (parentEntry) {
          parentEntry.recent_replies.push(entry);
        }
      }
    },
    'getEntries': function() {
      return entries;
    },
    'json': function() {
      return data;
    }
  };

  return that;
};

/**
 * Mock a Canvas discussion entry
 *
 * @param {CanvasUser}    user          The user who created the discussion entry
 * @param {Number}        [parentId]    The id of the parent entry in case this is a reply
 */
var CanvasDiscussionEntry = module.exports.CanvasDiscussionEntry = function(user, parentId) {
  return {
    'created_at': '2015-06-08T18:38:26Z',
    'id': _.random(100000),
    'parent_id': parentId,
    'updated_at': '2015-06-08T18:38:26Z',
    'user_id': user.id,
    'user_name': user.fullName,
    'message': '<p>bazinga</p>',
    'read_state': 'read',
    'forced_read_state': false,
    'recent_replies': []
  };
};
