// col-categories/lib/api.js
/**
 * Copyright 2015 UC Berkeley (UCB) Licensed under the
 * Educational Community License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may
 * obtain a copy of the License at
 *
 *     http://opensource.org/licenses/ECL-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an "AS IS"
 * BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

var Joi = require('joi');

var DB = require('col-core/lib/db');
var log = require('col-core/lib/logger')('col-categories');

/**
 * Get a category
 *
 * @param  {Context}        ctx                         Standard context containing the current user and the current course
 * @param  {Number}         id                          The id of the category that should be retrieved
 * @param  {Function}       callback                    Standard callback function
 * @param  {Object}         callback.err                An error that occurred, if any
 * @param  {Category}       callback.category           The retrieved category
 * @api private
 */
var getCategory = module.exports.getCategory = function(ctx, id, callback) {
  var options = {
    'where': {
      'course_id': ctx.course.id,
      'id': id
    }
  };

  DB.Category.find(options).complete(function(err, category) {
    if (err) {
      log.error({'err': err, 'id': id}, 'Failed to retrieve a category');
      return callback({'code': 500, 'msg': err.message});
    } else if (!category) {
      log.debug({'err': err, 'id': id}, 'The category could not be found');
      return callback({'code': 404, 'msg': 'The category could not be found'});
    }

    return callback(null, category);
  });
};

/**
 * Get the categories for the current course
 *
 * @param  {Context}        ctx                         Standard context containing the current user and the current course
 * @param  {Function}       callback                    Standard callback function
 * @param  {Object}         callback.err                An error that occurred, if any
 * @param  {Category[]}     callback.categories         The categories in the current course
 */
var getCategories = module.exports.getCategories = function(ctx, callback) {
  // Get the categories from the DB. As we need to include the number of assets per category,
  // a raw query is used
  var sqlQuery = 'SELECT category.*, (SELECT COUNT(*)::int FROM assets_categories WHERE category_id = category.id) AS asset_count';
  sqlQuery +=    ' FROM categories AS category';
  sqlQuery +=    ' WHERE category.course_id = ?';
  var replacements = [ctx.course.id];

  DB.getSequelize().query(sqlQuery, DB.Category, {'replacements': replacements}).complete(function(err, results) {
    if (err) {
      log.error({'err': err, 'course': ctx.course}, 'Failed to get the categories in the current course');
      return callback({'code': 500, 'msg': err.message});
    }

    return callback(null, results);
  });
};

/**
 * Get a set of categories by their id
 *
 * @param  {Context}        ctx                         Standard context containing the current user and the current course
 * @param  {Number[]}       ids                         The ids of the categories that should be retrieved
 * @param  {Function}       callback                    Standard callback function
 * @param  {Object}         callback.err                An error that occurred, if any
 * @param  {Category[]}     callback.categories         The retrieved categories
 */
var getCategoriesById = module.exports.getCategoriesById = function(ctx, ids, callback) {
  // Return immediately if no category ids have been provided
  if (!ids || ids.length === 0) {
    return callback(null, []);
  }

  // Parameter validation
  var validationSchema = Joi.object().keys({
    'ids': Joi.array().unique().items(Joi.number())
  });

  var validationResult = Joi.validate({
    'ids': ids
  }, validationSchema);

  if (validationResult.error) {
    return callback({'code': 400, 'msg': validationResult.error.details[0].message});
  }

  // Get the categories from the DB
  var options = {
    'where': {
      'course_id': ctx.course.id,
      'id': ids
    }
  };

  DB.Category.findAll(options).complete(function(err, categories) {
    if (err) {
      log.error({'err': err, 'ids': ids}, 'Failed to get a set of categories');
      return callback({'code': 500, 'msg': err.message});
    } else if (categories.length !== ids.length) {
      log.error({'err': err, 'ids': ids}, 'Could not retrieve all requested categories');
      return callback({'code': 404, 'msg': 'Could not retrieve all requested categories'});
    }

    return callback(null, categories);
  });
};

/**
 * Create a new category
 *
 * @param  {Context}        ctx                         Standard context containing the current user and the current course
 * @param  {String}         title                       The name of the category
 * @param  {Number}         [canvasAssignmentId]        The id of the assignment to which this category is linked
 * @param  {Function}       callback                    Standard callback function
 * @param  {Object}         callback.err                An error that occurred, if any
 * @param  {Category}       callback.category           The created category
 */
var createCategory = module.exports.createCategory = function(ctx, title, canvasAssignmentId, callback) {
  // Parameter validation
  var validationSchema = Joi.object().keys({
    'title': Joi.string().max(255).required(),
    'canvasAssignmentId': Joi.number().optional()
  });

  var validationResult = Joi.validate({
    'title': title,
    'canvasAssignmentId': canvasAssignmentId
  }, validationSchema);

  if (validationResult.error) {
    return callback({'code': 400, 'msg': validationResult.error.details[0].message});
  }

  // Only course administrators are allowed to create new categories
  if (!ctx.user.is_admin) {
    log.error({'course': ctx.course}, 'Unauthorized to create a new category');
    return callback({'code': 401, 'msg': 'Unauthorized to create a new category'});
  }

  // Create the category in the DB
  var category = {
    'course_id': ctx.course.id,
    'title': title,
    'canvas_assignment_id': canvasAssignmentId
  };

  DB.Category.create(category).complete(function(err, category) {
    if (err) {
      log.error({'err': err}, 'Failed to create a new category');
      return callback({'code': 500, 'msg': err.message});
    }

    // Add the asset count to the category
    category = category.toJSON();
    category.asset_count = 0;

    return callback(null, category);
  });
};

/**
 * Edit a category
 *
 * @param  {Context}        ctx                             Standard context containing the current user and the current course
 * @param  {Number}         id                              The id of the category that is being edited
 * @param  {String}         title                           The updated category name
 * @param  {Function}       callback                        Standard callback function
 * @param  {Object}         callback.err                    An error that occurred, if any
 * @param  {Category}       callback.category               The updated category
 */
var editCategory = module.exports.editCategory = function(ctx, id, title, callback) {
  // Parameter validation
  var validationSchema = Joi.object().keys({
    'id': Joi.number().required(),
    'title': Joi.string().max(255).required()
  });

  var validationResult = Joi.validate({
    'id': id,
    'title': title
  }, validationSchema);

  if (validationResult.error) {
    return callback({'code': 400, 'msg': validationResult.error.details[0].message});
  }

  // Only course administrators are allowed to edit a category
  if (!ctx.user.is_admin) {
    log.error({'id': id}, 'Unauthorized to edit a category');
    return callback({'code': 401, 'msg': 'Unauthorized to edit a category'});
  }

  // Retrieve the category that is being edited
  getCategory(ctx, id, function(err, category) {
    if (err) {
      return callback(err);
    }

    // Update the category in the DB
    var update = {
      'title': title
    };

    category.updateAttributes(update).complete(function(err, category) {
      if (err) {
        log.error({'err': err, 'id': id}, 'Failed to update a category');
        return callback({'code': 500, 'msg': err.message});
      }

      // Retrieve the number of assets associated to the category
      var sqlQuery = 'SELECT COUNT(*)::int AS count FROM assets_categories WHERE category_id = ?';
      var replacements = [category.id];
      DB.getSequelize().query(sqlQuery, {'replacements': replacements}).complete(function(err, result) {
        if (err) {
          log.error({'err': err, 'category': category.id}, 'Failed to get the number of assets associated to a category');
          return callback({'code': 500, 'msg': err.message});
        }

        // Add the asset count to the category
        category = category.toJSON();
        category.asset_count = result[0][0].count;

        return callback(null, category);
      });
    });
  });
};

/**
 * Delete a category
 *
 * @param  {Context}        ctx                             Standard context containing the current user and the current course
 * @param  {Number}         id                              The id of the category that is being deleted
 * @param  {Function}       callback                        Standard callback function
 * @param  {Object}         callback.err                    An error that occurred, if any
 */
var deleteCategory = module.exports.deleteCategory = function(ctx, id, callback) {
  // Parameter validation
  var validationSchema = Joi.object().keys({
    'id': Joi.number().required()
  });

  var validationResult = Joi.validate({
    'id': id
  }, validationSchema);

  if (validationResult.error) {
    return callback({'code': 400, 'msg': validationResult.error.details[0].message});
  }

  // Only course administrators are allowed to delete a category
  if (!ctx.user.is_admin) {
    log.error({'id': id}, 'Unauthorized to delete a category');
    return callback({'code': 401, 'msg': 'Unauthorized to delete a category'});
  }

  // Retrieve the category that is being deleted
  getCategory(ctx, id, function(err, category) {
    if (err) {
      return callback(err);
    }

    // Delete the category from the DB
    category.destroy().complete(function(err) {
      if (err) {
        log.error({'err': err, 'category': category.id}, 'Failed to delete a category');
        return callback({'code': 500, 'msg': err.message});
      }

      return callback();
    });
  });
};
